File: bench/java/src/main/java/ming/Tokenizer.java
package ming;

import java.util.ArrayList;
import java.util.List;

public class Tokenizer {
    public enum TokenType {
        LPAREN, RPAREN, SYMBOL, INTEGER, BOOLEAN, STRING, EOF
    }

    public record Token(TokenType type, String value, int pos) {}

    private final String input;
    private int pos;

    public Tokenizer(String input) {
        this.input = input;
        this.pos = 0;
    }

    public List<Token> tokenize() throws EvalError {
        List<Token> tokens = new ArrayList<>();
        while (pos < input.length()) {
            skipWhitespaceAndComments();
            if (pos >= input.length()) break;

            char c = input.charAt(pos);
            if (c == '(') {
                tokens.add(new Token(TokenType.LPAREN, "(", pos));
                pos++;
            } else if (c == ')') {
                tokens.add(new Token(TokenType.RPAREN, ")", pos));
                pos++;
            } else if (c == '"') {
                tokens.add(readString());
            } else if (c == '#') {
                tokens.add(readHash());
            } else {
                tokens.add(readAtom());
            }
        }
        tokens.add(new Token(TokenType.EOF, "", pos));
        return tokens;
    }

    private void skipWhitespaceAndComments() {
        while (pos < input.length()) {
            char c = input.charAt(pos);
            if (Character.isWhitespace(c)) {
                pos++;
            } else if (c == ';') {
                while (pos < input.length() && input.charAt(pos) != '\n') pos++;
            } else {
                break;
            }
        }
    }

    private Token readString() throws EvalError {
        int start = pos;
        pos++; // skip opening quote
        StringBuilder sb = new StringBuilder();
        while (pos < input.length() && input.charAt(pos) != '"') {
            if (input.charAt(pos) == '\\') {
                pos++;
                if (pos >= input.length()) throw new EvalError("Unterminated string");
                char esc = input.charAt(pos);
                switch (esc) {
                    case 'n' -> sb.append('\n');
                    case 't' -> sb.append('\t');
                    case '\\' -> sb.append('\\');
                    case '"' -> sb.append('"');
                    default -> { sb.append('\\'); sb.append(esc); }
                }
            } else {
                sb.append(input.charAt(pos));
            }
            pos++;
        }
        if (pos >= input.length()) throw new EvalError("Unterminated string");
        pos++; // skip closing quote
        return new Token(TokenType.STRING, sb.toString(), start);
    }

    private Token readHash() throws EvalError {
        int start = pos;
        pos++; // skip #
        if (pos >= input.length()) throw new EvalError("Unexpected end after #");
        char c = input.charAt(pos);
        if (c == 't') {
            pos++;
            // Check it's not part of a longer symbol like #true
            if (pos < input.length() && !isDelimiter(input.charAt(pos))) {
                // read rest
                StringBuilder sb = new StringBuilder("#t");
                while (pos < input.length() && !isDelimiter(input.charAt(pos))) {
                    sb.append(input.charAt(pos));
                    pos++;
                }
                if (sb.toString().equals("#true")) return new Token(TokenType.BOOLEAN, "true", start);
                throw new EvalError("Unknown literal: " + sb);
            }
            return new Token(TokenType.BOOLEAN, "true", start);
        } else if (c == 'f') {
            pos++;
            if (pos < input.length() && !isDelimiter(input.charAt(pos))) {
                StringBuilder sb = new StringBuilder("#f");
                while (pos < input.length() && !isDelimiter(input.charAt(pos))) {
                    sb.append(input.charAt(pos));
                    pos++;
                }
                if (sb.toString().equals("#false")) return new Token(TokenType.BOOLEAN, "false", start);
                throw new EvalError("Unknown literal: " + sb);
            }
            return new Token(TokenType.BOOLEAN, "false", start);
        }
        throw new EvalError("Unknown # literal at position " + start);
    }

    private Token readAtom() {
        int start = pos;
        StringBuilder sb = new StringBuilder();
        while (pos < input.length() && !isDelimiter(input.charAt(pos))) {
            sb.append(input.charAt(pos));
            pos++;
        }
        String val = sb.toString();
        // Check if integer
        try {
            Long.parseLong(val);
            return new Token(TokenType.INTEGER, val, start);
        } catch (NumberFormatException e) {
            return new Token(TokenType.SYMBOL, val, start);
        }
    }

    private boolean isDelimiter(char c) {
        return Character.isWhitespace(c) || c == '(' || c == ')' || c == '"' || c == ';';
    }
}

File: bench/scala/src/main/scala/ming/Tokenizer.scala
package ming

enum Token:
  case LParen
  case RParen
  case Str(value: String)
  case Atom(value: String)

object Tokenizer:

  def tokenize(input: String): List[Token] =
    val tokens = scala.collection.mutable.ListBuffer[Token]()
    var i      = 0
    while i < input.length do i = readToken(input, i, tokens)
    tokens.toList

  private def readToken(
    input: String,
    pos: Int,
    tokens: scala.collection.mutable.ListBuffer[Token]
  ): Int =
    val ch = input(pos)
    ch match
      case _ if ch.isWhitespace => pos + 1
      case ';' =>
        var i = pos
        while i < input.length && input(i) != '\n' do i += 1
        i
      case '(' =>
        tokens += Token.LParen
        pos + 1
      case ')' =>
        tokens += Token.RParen
        pos + 1
      case '"' => readString(input, pos + 1, tokens)
      case '\'' =>
        tokens += Token.Atom("quote-sugar")
        pos + 1
      case _ =>
        val sb = new StringBuilder
        var i  = pos
        while i < input.length && !input(i).isWhitespace && input(i) != '(' && input(i) != ')' && input(
            i
          ) != ';' && input(i) != '"'
        do
          sb += input(i)
          i += 1
        tokens += Token.Atom(sb.toString)
        i

  private def readString(
    input: String,
    start: Int,
    tokens: scala.collection.mutable.ListBuffer[Token]
  ): Int =
    val sb = new StringBuilder
    var i  = start
    while i < input.length && input(i) != '"' do
      if input(i) == '\\' && i + 1 < input.length then
        input(i + 1) match
          case 'n'   => sb += '\n'; i += 2
          case 't'   => sb += '\t'; i += 2
          case '\\'  => sb += '\\'; i += 2
          case '"'   => sb += '"'; i += 2
          case other => sb += '\\'; sb += other; i += 2
      else
        sb += input(i)
        i += 1
    if i < input.length then i += 1 // skip closing "
    tokens += Token.Str(sb.toString)
    i
